### PR TITLE
Fix transformer masks usage

### PR DESCRIPTION
**Pull request by EmmaRenauld**
Fixes Issue #92554 

The float usage in masks for the Transformer is considered deprecated, but still accepted.

 - `key_padding_mask` was ignored in the Transformer's multihead attention in the case where user creates a float mask. Fixing the float case.
 - Transformer's function `generate_square_subsequent_mask` generates float masks by default. Changing to bool mask.
 - Also adding a new `generate_padding_mask` in the Transformer, to help user create correct masks.

Don't hesitate to correct me for the computation efficiency.

### CODE REVIEW

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -151,11 +151,40 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
...
-        r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
-            Unmasked positions are filled with float(0.0).
+    def generate_square_subsequent_mask(sz: int, device='cpu', use_bool_type: bool = True) -> Tensor:
+        r"""Generate a square mask for the sequence. The masked positions are filled with True, and
+        unmasked positions with False.
+            With use_bool_type = False, reverts to deprecated float usage: masked positions are filled
+        with float('-inf'), and unmasked positions with float(0.0).
         """
-        return torch.triu(torch.full((sz, sz), float('-inf'), device=device), diagonal=1)
+        mask = torch.triu(torch.full((sz, sz), float('-inf'), device=device), diagonal=1)
```
Comment (vadimkantorov) [inline torch/nn/modules/transformer.py:160]: should this just pass directly a correct `dtype=` / fill value to `torch.full`?

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -151,11 +151,40 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
...
+
+        Args:
+            unpadded_lengths: List of each sequence's length in the batch.
+            max_len: The length of sequences after padding.
+
+        """
+        batch_size = len(unpadded_lengths)
+        fill_value = float(-inf)
+
+        mask = torch.zeros(batch_size, max_len)
```
Comment (vadimkantorov) [inline torch/nn/modules/transformer.py:181]: Also, this can be done without the loop by torch.arange + expand + lt I think

**Comment by EmmaRenauld** [comment]
Sorry. Someone else had already placed a PR. See #92733
